src/vcf/record.cpp

```diff
@@ -620,28 +661,50 @@ namespace ebi
+    {
+        bool valid;
+        long cardinality;
+        size_t ploidy = 0;
+        if (number != G) {
+            check_sample_field_cardinality(field, values, number, ploidy, valid, cardinality);
+        } else {
+            // this case is not well defined, we just skip this check and allow any cardinality
+            // for further discussion see https://github.com/samtools/hts-specs/issues/272
+        }
+    }
+
+    void Record::check_sample_field_cardinality(std::string const &field, std::vector<std::string> const &values,
+                                                std::string const &number, size_t ploidy, bool &valid,
+                                                long &cardinality) const
     {
-        long expected;
-        if(not is_valid_cardinality(number, alternate_alleles.size(), expected)) {
-            raise(std::make_shared<Error>(line, field + " meta specification Number=" + number + " is not one of [A, R, G, ., <non-negative number>]"));
+        valid = is_valid_cardinality(number, alternate_alleles.size(), ploidy, cardinality);
+        if (not valid) {
+            raise(std::make_shared<Error>(line, " meta specification Number=" + number
+                    + " is not one of [A, R, G, ., <non-negative number>]"));
         }
 
         bool number_matches = true;
-        if (expected > 0) {
-            // The number of values must match the expected
-            number_matches = (values.size() == static_cast<size_t>(expected));
-        } else if (expected == 0) {
+        if (cardinality > 0) {
+            // The number of values must match the expected cardinality
+            number_matches = (values.size() == static_cast<size_t>(cardinality));
+        } else if (cardinality == 0) {
             // There will be one empty value that needs to be specifically checked
             number_matches = values.size() == 0 || values.size() == 1;
         } else {
-            // if number=".", then `expected` was set to -1, and it should always match, letting `number_matches` as true
+            // if number=".", then `cardinality` was set to -1, and it should always match, letting `number_matches` as true
         }
 
         if (!number_matches) {
-            raise(std::make_shared<Error>(line, " specification Number=" + number + " (expected " + std::to_string(expected) + " value(s))"));
+            std::string detailed_message;
+            if (number == G) {
+                detailed_message = ". It must derive its number of values from the ploidy of GT (if present), or "
+                        "assume a diploidy. Contains " + std::to_string(values.size()) + " value(s), expected "
+                        + std::to_string(cardinality) + " (derived from ploidy " + std::to_string(ploidy) + ")";
+            }
+            raise(std::make_shared<Error>(line, " specification Number=" + number + " (expected "
+                    + std::to_string(cardinality) + " value(s))", detailed_message));
```

Comment:
I think it would be more clear if the first part of the message was set first as a separate variable, then the detail in the other one. What does this concatenate if `detailed_message` is undefined?

Reply:
agree, but detailed_message can not be undefined, it's just an empty string.